### DISCUSSION TITLE
Expand the Coding Styleguide

### DISCUSSION THREAD

**Discussion by thebonbon** [EveronLife/EveronLife#51] [Ideas]
The Coding Styleguide should be expanded a little to set some basics (Some are in the conventions link but should also be written here).

**Script Prefix**
Some scripts already use this but I have not seen it beeing mentioned anywhere.
Every EveronLife Script should have the `EL_` prefix in the name.

`EL_TraderManager.c`

> Bohemia Interactive scripts have been prefixed with SCR_.

**Component Category**
A category should also be added to new components / entities to keep things organzied.
I suggest `"EveronLife/Game/FEATURE"`.
If there are a lot of components a subfeature could be added  `"EveronLife/Game/FEATURE/SUBFEATURE"`

This needs to be added in the first line e.g:
`[ComponentEditorProps(category: "EveronLife/Game/Farming")]`
or
`[ComponentEditorProps(category: "EveronLife/Game/Farming/Managers")]`

**Method (From https://community.bistudio.com/wiki/Arma_Reforger:Scripting:_Conventions)**

> All methods must be separated using this sequence of characters: two slashes followed by 96 dashes (https://community.bistudio.com/wiki/Arma_Reforger:Scripting:_Conventions#Example))
> `//------------------------------------------------------------------------------------------------`

This is currently not used in any script.

**Code Comments**

> Code comments are surprisingly not a must-have; code organisation combined to variable names should be enough to be read by a human, then comment can be used:
> 
> a comment should explain why the code is written this way
> a comment should not tell what the code does; code should be self-explanatory
> as a last resort in the event of a complex piece of code, a comment can be used to describe what the code actually does - or at least its intention
> On the other hand, Documentation is more than welcome as it provides information from the outside without having to read the code. Enfusion uses [Doxygen](https://www.doxygen.nl/manual/docblocks.html) format.

Since this is a framework we should comment most things.
e.g 
```
//------------------------------------------------------------------------------------------------
//! Compare two integers, return the larger one.
//\param a First parameter to be compared with the second one
//\param b Second parameter to be compared with the first one
//\return Returns true if a is equal to b, false otherwise.
bool IsEqual(int a, int b)
```

**Curly Braces {}**
I have seen both beeing used in the scripts. We need to decide on one.
Every Reforger script uses the first one. I suggest sticking to that.
```
Foo() 
{
}

Foo() {
}
```

*Edit
**Folder Structure**
In addition to that a proper folder structure should be implemented.
Scripts Folder:
Right now we have two main folders for scripting systems: "Components" and "Features"
The Features folder also contains actions and a component.

![image](https://user-images.githubusercontent.com/45827986/171004514-a6c742b7-893e-4025-a584-5351a79d6343.png)
These two folders should be merged / organized better

**Comment by thebonbon** [comment]
For Curly Braces:

I overlooked this from the scripting conventions:
> Curly braces must always be on a new line - Enforce Script uses [Allman style](https://en.wikipedia.org/wiki/Indentation_style#Allman_style)

**Comment by Arkensor** [comment]
> Script Prefix
Some scripts already use this but I have not seen it beeing mentioned anywhere.
Every EveronLife Script should have the EL_ prefix in the name.

Yes, everything will need to be prefixed. Our latest findings regarding how the resource names are generated lead us to believe we need to even prefix assets like .conf .fbx etc. They clash if a different mod happens to have the same file name. e.g. `EveronLife/blah.txt` is the same as `Rush/blah.txt`. So non-prefixed files indicate an intentional override in the directories meanwhile `EL_` content is kept safe and sound from any conflicts. Not sure yet if we want to do this for non-scripts. It looks kinda ugly to prefix all the files :/

> Component Category
A category should also be added to new components / entities to keep things organzied.
I suggest "EveronLife/Game/FEATURE".
If there are a lot of components a subfeature could be added "EveronLife/Game/FEATURE/SUBFEATURE"

I generally agree with having component categories. We just did not take a look at those yet.

> All methods must be separated using this sequence of characters: two slashes followed by 96 dashes ([community.bistudio.com/wiki/Arma_Reforger:Scripting:_Conventions#Example](https://community.bistudio.com/wiki/Arma_Reforger:Scripting:_Conventions#Example)))
//------------------------------------------------------------------------------------------------

That is an odd one. I have no idea why they do that tbh. I suppose it is because you can not collapse code. I mean we can refer to that for now. If people complain about it, it would be easy to mass remove down the road.

> Curly Braces {}
I have seen both beeing used in the scripts. We need to decide on one.
Every Reforger script uses the first one. I suggest sticking to that.

100%. The code on the repo right now was written before we sat down to make the guidelines for this project. { goes on a new line, no debate. 

> Folder Structure
In addition to that a proper folder structure should be implemented.

This is what I am not sure about yet. I tried guessing into the future to see what requirements we might have. I liked how in Altis Life you had one folder per "feature" that contained a lot of the related files. But AR splits models from prefabs from scripts from configs from world files etc. So a vertical slice approach might not work too well. 

What value do you see in splitting off components? I would think a lot of them will belong only to one "feature". We might have some shared ones but those could be "features" depending on other "features". And they are put where they belong primarily. Having the same sub folders for components and then for features would not prove too much value. But maybe you had other reasons for this?

**Reply by thebonbon** [reply]
> This is what I am not sure about yet. I tried guessing into the future to see what requirements we might have. I liked how in Altis Life you had one folder per "feature" that contained a lot of the related files. But AR splits models from prefabs from scripts from configs from world files etc. So a vertical slice approach might not work too well.
> 
> What value do you see in splitting off components? I would think a lot of them will belong only to one "feature". We might have some shared ones but those could be "features" depending on other "features". And they are put where they belong primarily. Having the same sub folders for components and then for features would not prove too much value. But maybe you had other reasons for this?

I was just confused why there is a component script in the features folder (EL_LoadoutManagerComponent.c).

So all components from "X system" go into the `Scripts/Components/XSystem`
And anything else related to "X system" that is not a component goes into `Scripts/Features/X`

**Reply by Arkensor** [reply]
> So all components from "X system" go into the Scripts/Components/XSystem
And anything else related to "X system" that is not a component goes into Scripts/Features/X

What is the difference or advantage of putting the components into a separate folder instead of with the rest of the feature?
Perhaps you thought we wanted to separate them because of the few existing ones? Those were just because the game happened to automatically put them there. As far as I can tell from the vanilla files they often put together logic and components e.g. `ArmaReforger:scripts/GameCode/Vehicle`. I think we should keep the code together until we find a good reason to separate out components.

**Reply by thebonbon** [reply]
No advantage thats just how I saw the vanilla files. I only looked in `scripts/Game/`  `scripts/Game/Components ` There are a lot of Components of different systems in this folder.
But I guess a single features folder would be best.

**Comment by thebonbon** [comment]
These Styleguides need to be added to a linter to enforce.
Linter discussion https://github.com/EveronLife/EveronLife/discussions/64

**Comment by Arkensor** [comment]
One thing we need to add to my deepest regret is that as long as it breaks autocompletion inside the workbench do not use the `auto` keyword.

**Comment by Arkensor** [comment]
I would like to get feedback on this code I just read in one of our PRs:
```cs
void GoBack(int steps)
{
	int nextIdx = m_EntryIndex - steps;
	if(nextIdx <= 0) Reset();
	else m_EntryIndex = nextIdx;
}
```
As you probably guessed while reading I am wondering about how we deal with single statements, especially here with this if. The code is still understandable and of course shorter than 
```cs
	if(nextIdx <= 0)
	{
		Reset();
	}
	else
	{
		m_EntryIndex = nextIdx;
	}
```
However, I think the readability of the version with code blocks is better and it is easier to place breakpoints and add code later. That is the reason I personally add a code block to if statements or while etc even if it's just one statement. Exception's from the rule are only things like `if(xxx) return;` that you usually never expand or never debug the statement coming after the if.

I would like to get your opinions if you think that we should enforce any rule regarding this or leave it up to the dev who is working on it if he wants to wrap single-line statements or not?

**Reply by Jaccodouma** [reply]
I personally like things like `if (nextIdx <= 0) Reset()` on one line, but as soon as there's an else it becomes messy to have it as the first example imo. I think it's mainly readablilty, I already almost read over the else in your top example and assumed you were creating another variable.

**Reply by BenRauzi** [reply]
Ditto to @Jaccodouma from my perspective.

As soon as you have else I'd recommend code blocks. I don't mind either way though as long as `if (xxx) return;` is fine

**Reply by thebonbon** [reply]
For single line IFs I always do it in the next line but without { }
```
if(nextIdx <= 0)
    Reset();
else
    m_EntryIndex = nextIdx;
```
